crates/bevy_ecs/src/world/mod.rs

```diff
@@ -2635,6 +2635,212 @@ impl World {
+    ///
+    /// # Errors
+    ///
+    /// See [`GetComponentReflectError`](error::GetComponentReflectError) for the possible errors and their descriptions.
+    ///
+    /// # Example
+    ///
+    /// See the documentation for [`World::get_reflect()`].
+    ///
+    /// # Note
+    /// Requires the feature `bevy_reflect` (included in the default features).
+    #[inline]
+    #[cfg(feature = "bevy_reflect")]
+    pub fn get_reflect_mut(
+        &mut self,
+        entity: Entity,
+        type_id: TypeId,
+    ) -> Result<Mut<'_, dyn bevy_reflect::Reflect>, error::GetComponentReflectError> {
+        use bevy_reflect::ReflectFromPtr;
+
+        use crate::prelude::AppTypeRegistry;
+
+        use error::GetComponentReflectError;
+
+        // little clone() + read() dance so we a) don't keep a borrow of `self` and b) don't drop a
+        // temporary (from read()) too  early.
+        let Some(app_type_registry) = self.get_resource::<AppTypeRegistry>().cloned() else {
+            return Err(GetComponentReflectError::MissingAppTypeRegistry);
+        };
+        let type_registry = app_type_registry.read();
+
+        let Some(reflect_from_ptr) = type_registry.get_type_data::<ReflectFromPtr>(type_id) else {
+            return Err(GetComponentReflectError::MissingReflectFromPtrTypeData(
+                type_id,
+            ));
+        };
+
+        let Some(component_id) = self.components().get_id(type_id) else {
+            return Err(GetComponentReflectError::NoCorrespondingComponentId(
+                type_id,
+            ));
+        };
+
+        // HACK: Only required for the `None`-case/`else`-branch, but it borrows `self`, which will
+        // already be borrowed by `self.get_mut_by_id()`, and I didn't find a way around it.
+        let component_name = self
+            .components()
+            .get_name(component_id)
+            .map(ToString::to_string);
+
+        let Some(comp_mut_untyped) = self.get_mut_by_id(entity, component_id) else {
+            return Err(GetComponentReflectError::EntityDoesNotHaveComponent {
+                entity,
+                type_id,
+                component_id,
+                component_name,
+            });
+        };
+
+        // let app_type_registry = self.get_resource::<AppTypeRegistry>()?.clone();
+        // let type_registry = app_type_registry.read();
+        // let reflect_from_ptr = type_registry.get_type_data::<ReflectFromPtr>(type_id)?;
+        //
+        // let component_id = self.components().get_id(type_id)?;
+        // let comp_mut_untyped = self.get_mut_by_id(entity, component_id)?;
+
+        // SAFETY:
+        // - `comp_mut_untyped` is guaranteed to point to an object of type `type_id`
+        // - `reflect_from_ptr` was constructed for type `type_id`
+        // - Assertion that checks this equality is present
+        let comp_mut_typed = comp_mut_untyped.map_unchanged(|ptr_mut| unsafe {
+            assert_eq!(
```

Comment:
Same here, should this be a `debug_assert_eq!`?

Reply:
Guess it should stay as an `assert_eq!` for now since it's required for soundness, at least according to the other discussion.

Reply:
Taking the :+1: as agreement :)